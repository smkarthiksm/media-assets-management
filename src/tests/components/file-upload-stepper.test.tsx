import '@testing-library/jest-dom';
import { renderWithProviders } from '../test-utility';
import { fireEvent, screen } from '@testing-library/react';
import FileUploadStepperComponent from '../../components/file-upload-stepper/file-upload-stepper';
import sampleAudioFile from '../sample-files-for-testing/sample-audio.mp3';
import userEvent from '@testing-library/user-event';
import { allFilesStateStub } from '../stubs/all-files-state.stub';

describe('FileUploadStepperComponent', () => {
  test('should render', () => {
    const { container } = renderWithProviders(<FileUploadStepperComponent />);
    expect(container).toMatchSnapshot();
  });

  test('should temporarily store all the files in the state that were selected in the file uploader', async () => {
    const { container, store } = renderWithProviders(
      <FileUploadStepperComponent />,
    );
    // mock DOM createElement API to mock eventListener
    global.URL.createObjectURL = jest.fn(() => 'random_id');
    const mockAudioElement = document.createElement('audio');
    mockAudioElement.addEventListener = (type: string, handler: any) =>
      handler();

    const mockFunction = jest
      .spyOn(document, 'createElement')
      .mockReturnValue(mockAudioElement);

    const files = [
      new File([sampleAudioFile], 'sample-file-1.mp3', {
        type: 'audio/mp3',
      }),
      new File([sampleAudioFile], 'sample-file-2.mp3', {
        type: 'audio/mp3',
      }),
    ];
    const fileInput = container.getElementsByClassName(
      'file-input',
    )[0] as HTMLInputElement;

    expect(store.getState().fileUploadStepper.files).toHaveLength(0);
    await userEvent.upload(fileInput, files);
    expect(store.getState().fileUploadStepper.files).toHaveLength(2);
    mockFunction.mockRestore();
  });

  test(`should update first file's text fields in the next step after files are selected`, async () => {
    const { store } = renderWithProviders(<FileUploadStepperComponent />, {
      preloadedState: {
        fileUploadStepper: {
          activeStep: 1,
          files: allFilesStateStub.allFiles,
        },
      },
    });
    const titleTextField = screen.getAllByRole('textbox', { name: 'Title' })[0];
    const albumTextField = screen.getAllByRole('textbox', { name: 'Album' })[0];
    const artistTextField = screen.getAllByRole('textbox', {
      name: 'Artist',
    })[0];
    expect(store.getState().fileUploadStepper.files[0].title).toEqual(
      'testTitle1',
    );
    expect(store.getState().fileUploadStepper.files[0].album).toEqual(
      'testAlbum1',
    );
    expect(store.getState().fileUploadStepper.files[0].artist).toEqual(
      'testArtist1',
    );
    fireEvent.change(titleTextField, {
      target: { value: 'Updated new file title' },
    });
    fireEvent.change(albumTextField, {
      target: { value: 'Updated new file album' },
    });
    fireEvent.change(artistTextField, {
      target: { value: 'Updated new file artist' },
    });
    expect(store.getState().fileUploadStepper.files[0].title).toEqual(
      'Updated new file title',
    );
    expect(store.getState().fileUploadStepper.files[0].album).toEqual(
      'Updated new file album',
    );
    expect(store.getState().fileUploadStepper.files[0].artist).toEqual(
      'Updated new file artist',
    );
  });

  test(`should remove the second file from the files selected to upload`, async () => {
    const { store } = renderWithProviders(<FileUploadStepperComponent />, {
      preloadedState: {
        fileUploadStepper: {
          activeStep: 1,
          files: allFilesStateStub.allFiles,
        },
      },
    });
    const deleteButton = screen.getAllByRole('button', { name: 'Remove' })[1];
    expect(store.getState().fileUploadStepper.files).toHaveLength(2);
    fireEvent.click(deleteButton);
    expect(store.getState().fileUploadStepper.files).toHaveLength(1);
    expect(store.getState().fileUploadStepper.files).toEqual([
      allFilesStateStub.allFiles[0],
    ]);
  });

  test(`should go to first step if all files are removed that were selected to upload`, async () => {
    const { store } = renderWithProviders(<FileUploadStepperComponent />, {
      preloadedState: {
        fileUploadStepper: {
          activeStep: 1,
          files: [allFilesStateStub.allFiles[0]],
        },
      },
    });
    expect(store.getState().fileUploadStepper.files).toHaveLength(1);
    const deleteButton = screen.getAllByRole('button', { name: 'Remove' })[0];
    fireEvent.click(deleteButton);
    expect(store.getState().fileUploadStepper.files).toHaveLength(0);
    expect(store.getState().fileUploadStepper.activeStep).toEqual(0);
  });

  test(`should go to step 1 when back button is clicked`, async () => {
    const { store } = renderWithProviders(<FileUploadStepperComponent />, {
      preloadedState: {
        fileUploadStepper: {
          activeStep: 1,
          files: allFilesStateStub.allFiles,
        },
      },
    });
    const backButton = screen.getByRole('button', { name: 'Back' });
    expect(store.getState().fileUploadStepper.activeStep).toEqual(1);
    fireEvent.click(backButton);
    expect(store.getState().fileUploadStepper.activeStep).toEqual(0);
  });

  test(`should disable next button when no files are selected to upload`, async () => {
    renderWithProviders(<FileUploadStepperComponent />, {
      preloadedState: {
        fileUploadStepper: {
          activeStep: 0,
          files: [],
        },
      },
    });
    const nextButton = screen.getByRole('button', { name: 'Next' });
    expect(nextButton).toBeDisabled();
  });

  test(`should go to next step when files are selected to upload and next button is clicked`, async () => {
    const { store } = renderWithProviders(<FileUploadStepperComponent />, {
      preloadedState: {
        fileUploadStepper: {
          activeStep: 0,
          files: allFilesStateStub.allFiles,
        },
      },
    });
    expect(store.getState().fileUploadStepper.activeStep).toEqual(0);
    const nextButton = screen.getByRole('button', { name: 'Next' });
    fireEvent.click(nextButton);
    expect(store.getState().fileUploadStepper.activeStep).toEqual(1);
  });
});
