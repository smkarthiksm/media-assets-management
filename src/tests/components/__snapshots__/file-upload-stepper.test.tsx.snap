// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUploadStepperComponent should render 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container css-s5hy0i-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <div
        class="MuiStepper-root MuiStepper-horizontal css-m5vj9m-MuiStepper-root"
      >
        <div
          class="MuiStep-root MuiStep-horizontal css-1bw0nnu-MuiStep-root"
        >
          <span
            class="MuiStepLabel-root MuiStepLabel-horizontal css-ascpo7-MuiStepLabel-root"
          >
            <span
              class="MuiStepLabel-iconContainer Mui-active css-vnkopk-MuiStepLabel-iconContainer"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root Mui-active css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="12"
                />
                <text
                  class="MuiStepIcon-text css-117w1su-MuiStepIcon-text"
                  dominant-baseline="central"
                  text-anchor="middle"
                  x="12"
                  y="12"
                >
                  1
                </text>
              </svg>
            </span>
            <span
              class="MuiStepLabel-labelContainer css-1vyamtt-MuiStepLabel-labelContainer"
            >
              <span
                class="MuiStepLabel-label Mui-active css-1hv8oq8-MuiStepLabel-label"
              >
                Upload
              </span>
            </span>
          </span>
        </div>
        <div
          class="MuiStepConnector-root MuiStepConnector-horizontal Mui-disabled css-j5w0w9-MuiStepConnector-root"
        >
          <span
            class="MuiStepConnector-line MuiStepConnector-lineHorizontal css-z7uhs0-MuiStepConnector-line"
          />
        </div>
        <div
          class="MuiStep-root MuiStep-horizontal css-1bw0nnu-MuiStep-root"
        >
          <span
            class="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled css-ascpo7-MuiStepLabel-root"
          >
            <span
              class="MuiStepLabel-iconContainer Mui-disabled css-vnkopk-MuiStepLabel-iconContainer"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="12"
                />
                <text
                  class="MuiStepIcon-text css-117w1su-MuiStepIcon-text"
                  dominant-baseline="central"
                  text-anchor="middle"
                  x="12"
                  y="12"
                >
                  2
                </text>
              </svg>
            </span>
            <span
              class="MuiStepLabel-labelContainer css-1vyamtt-MuiStepLabel-labelContainer"
            >
              <span
                class="MuiStepLabel-label Mui-disabled css-1hv8oq8-MuiStepLabel-label"
              >
                Verify data and save
              </span>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item css-1mkcasw-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-12 upload-file-container css-1sd55v0-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-lptiby-MuiGrid-root"
        >
          <input
            accept=".mp3,.aac,.wav,.mp4,.mov,.avi"
            class="file-input"
            id="file"
            multiple=""
            title=" "
            type="file"
          />
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
          >
            Accepted formats - .mp3,.aac,.wav,.mp4,.mov,.avi
          </h6>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        0
         files selected
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item css-9pwih8-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      />
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;
